Departure time patches

Abby-Wheelis opened the pull request:
## Describe your changes
additions of charging_session_timer, end_of_profile_schedule, sim_speed, and departure_time to the iso15118/evcc_comm_session_handler.py file to initialize these values used in the schedule calculation that includes departure time
addition of EAmount and DepartureTime to the EVState class definition
changes in iso15118/controller/simulator.py which convert hardcoded EAmount and DepartureTime with variables in the EVEREST_EV_STATE and include calculating a power curve based on the requested departure time and publishes it via mqtt
additions to iso15118/evcc/states/iso15118_2_states.py to handle the end of the charging profile

## Issue ticket number and link
[#220](https://github.com/EVerest/EVerest/issues/220)

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

Opening PR as a draft to continue the conversation about the changes.

hikinggrass committed:
Use find_package and Python3_EXECUTABLE to work with python venv (#32)

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas committed:
Added a proper validation function to check price_level_schedule and absolute_price_schedule (#31)

According the iso15118-20 schema file the charger can either send no price information, absolute or price_level.

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
enable departure time in comm session handler

Abby-Wheelis committed:
enable departure time in comm session handler

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
enable payment method in python

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
enable payment method in python

Abby-Wheelis committed:
enable ev state

Abby-Wheelis committed:
enable ev state

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
enable iso15118 2 states

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
enable iso15118 2 states

Abby-Wheelis committed:
enable departure time in simulator

Abby-Wheelis committed:
enable departure time in simulator

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Revert "enable payment method in python"

This reverts commit 504624fbe18f127df9435bcb1ffd5d5673447558.

Abby-Wheelis committed:
Revert "enable payment method in python"

This reverts commit 504624fbe18f127df9435bcb1ffd5d5673447558.

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove unused AuthEnum import

Abby-Wheelis committed:
remove unused AuthEnum import

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas commented:
I will have a look next week

SebaLukas reviewed (commented):
I had a first look and what I immediately noticed.
- Please do not use prints, but the log functions accordingly.
- There must be an alternative that Josev does not publish/subscribe directly to mqtt `(everest_external/nodered/..)`. If we allow something like that here, I can no longer guarantee that the EV sim will work. Especially with different versions of NodeRed config jsons.

SebaLukas committed:
Refactor TLS 1.3 support for D20 Ev side (#36)

Adding vehicle certs for d20 tls 1.3 support, changed ENABLE_TLS_1_3 from a const to a variable

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
correct whitespace

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>

Abby-Wheelis committed:
correct whitespace

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
address unused imports

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
address unused imports

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>

Abby-Wheelis committed:
whitespace fixes

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>

Abby-Wheelis committed:
whitespace fixes

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
cut down on log statements, convert to logger.debug

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
cut down on log statements, convert to logger.debug

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>

Abby-Wheelis committed:
fix log string formatting

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>

Abby-Wheelis committed:
fix log string formatting

Signed-off-by: Abby Wheelis <abby.wheelis@gmail.com>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
don't read DT or EAmount from mqtt, read from state

Abby-Wheelis committed:
don't read DT or EAmount from mqtt, read from state

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix nits

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix nits

Abby-Wheelis committed:
hardcode algorithm choice for now

avoids improper use of MQTT
the charging station will probably choose this internally, so we would want to model it some other way, rather than the same way we do user input like EA and DT

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
hardcode algorithm choice for now

avoids improper use of MQTT
the charging station will probably choose this internally, so we would want to model it some other way, rather than the same way we do user input like EA and DT

Abby-Wheelis committed:
remove sim speed

setting aside this simulation-only factor for later reconsideration

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove sim speed

setting aside this simulation-only factor for later reconsideration

Abby-Wheelis committed:
update condition from is to ==

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
update condition from is to ==

Abby-Wheelis committed:
logging updates

Abby-Wheelis committed:
logging updates

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove unused input

Abby-Wheelis committed:
remove unused input

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix nits

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix nits

Abby-Wheelis committed:
address nits

Abby-Wheelis committed:
address nits

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix whitespace and remove unused variables

Abby-Wheelis committed:
fix whitespace and remove unused variables

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
unused var

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
unused var

Abby-Wheelis committed:
Merge branch 'everest' into departure-time-patches

Abby-Wheelis committed:
remove unused imports

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
remove unused imports

Abby-Wheelis committed:
Merge branch 'departure-time-patches' of https://github.com/Abby-Wheelis/ext-switchev-iso15118 into departure-time-patches

Abby-Wheelis commented:
> I had a first look and what I immediately noticed.
> 
> * Please do not use prints, but the log functions accordingly.
> * There must be an alternative that Josev does not publish/subscribe directly to mqtt `(everest_external/nodered/..)`. If we allow something like that here, I can no longer guarantee that the EV sim will work. Especially with different versions of NodeRed config jsons.

There are no more print statements, just log functions.
I have re-configured to not subscribe to MQTT directly, but using the state instead to store the variables. 

This is ready for another review.

SebaLukas commented on iso15118/evcc/controller/simulator.py:
```diff
@@ -610,13 +618,35 @@ async def process_sa_schedules_v2(
+            formatted_curve = formatCurveData(new_schedule)
+            # Then Re-Publish the chosen curve as the final selection
+            logger.debug(f"About to publish {str(formatted_curve)=}")
+            mqtt_publish.single("everest_external/nodered/{}/evcc/active_powercurve", str(formatted_curve), hostname="mqtt-server")
```
I would also like to replace `mqtt_publish` here. It would be best to define a `publish_extern()` function here, as has been done for the everest publish vars.
I'll have a look at this over the next few days to see if I can get it in. It is important to me that Josev is free of mqtt publish and subscribe.

Abby-Wheelis replied:
I have removed the publish. It was only used to display the curve in the UI, which is not critical, and users are likely not to care how the car charges, just that it is charged enough by the time they leave.

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,158 @@
+    # define the input for closed-loop simulation
+    inputCL=np.zeros(shape=(1,numberSamples))
+    inputCL[0,:]=xd*np.ones(numberSamples)
+    print(f"Created input array with {EAmount=} and {numberSamples=}")
```
`logger.debug` -> `print`

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,158 @@
+        schedule_arr = []
+        for i in range(0, len(curve_arr) - 2):
+            schedule_arr.append(make_entry(curve_arr[i][0], curve_arr[i][1], curve_arr[i+1][1]))
+        print('Done')
```
`logger.debug` -> `print`

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,158 @@
+    curve_schedule = sample_schedule(curve_schedule)
+
+    if (len(curve_schedule) <= 2): # Check to see if we're done...
+        print("Done with profile, defaulting to SECC Schedule")
```
`logger.debug` -> `print`

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,158 @@
+        return secc_schedule
+
+    curve_schedule = convert_tuple_schedule(curve_schedule)
+    print("Returning a curve schedule of:", curve_schedule)
```
`logger.debug` -> `print`

SebaLukas commented on iso15118/evcc/states/power_curve.py:
```diff
@@ -0,0 +1,158 @@
+            max_phases_in_use = None
+            )
+        )
+    print('Katie:', temp)
```
`logger.debug` -> `print`

SebaLukas reviewed (commented):
Have a look on my comments.
When the PR is really ready to be reviewed, I'll also look at the logic of how departure time and EAmount are used.

Abby-Wheelis committed:
exchange print statements for logger calls

Abby-Wheelis committed:
exchange print statements for logger calls

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
fix nits

Abby-Wheelis committed:
fix nits

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

SebaLukas committed:
Adding EV MCS serivces (#38)

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
Merge remote-tracking branch 'upstream/everest' into departure-time-patches

SebaLukas committed:
Adding ISO15118-20 pause/resume to the ev side (#37)

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
add patches to support departure time

Signed-off-by: Abby Wheelis <Abigail.Wheelis@nrel.gov>

Abby-Wheelis committed:
add patches to support departure time

Abby-Wheelis committed:
Merge branch 'departure-time-patch-based' into departure-time-patches

Abby-Wheelis committed:
Merge branch 'departure-time-patches' of https://github.com/Abby-Wheelis/ext-switchev-iso15118 into departure-time-patches

Abby-Wheelis commented:
I made too many mistakes trying to fix the signing for DCO, closing this in favor of #40
